app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -1,21 +1,62 @@
 class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  before_action :set_user_service
+
   def facebook
-    @user = User.from_omniauth(request.env["omniauth.auth"])
+    handle_auth 'Facebook'
+  end
+
+  def handle_auth(kind)
+    if @user.present?
+      # user exists
+      @user.access_token
+      # this will throw if @user is not activated
+      sign_in_and_redirect @user, event: :authentication
+    else
+      help_handle_auth(kind)
+    end
+  end
 
+  def help_handle_auth(kind)
+    # user does not exist
+    # user.services.create(service_attrs)
+    intent = request.env['omniauth.params']['intent']
+    @user = User.from_omniauth(auth)
     if @user.persisted?
-      sign_in_and_redirect @user, event: :authentication #this will throw if @user is not activated
-      set_flash_message(:notice, :success, kind: "Facebook") if is_navigational_format?
-      # @user.update token: request.env['omniauth.auth'].credentials.token
-      # @user.update friends: User.add_friends(@user.token)
+      if intent == 'sign_up'
+        ahoy.track 'New user sign up', name: @user.name
+        AddMoviesToUserWorker.perform_async(@user.id)
+        AddFriendsToUserWorker.perform_async(@user.id)
+      end
+      sign_in_and_redirect @user, event: :authentication # this will throw if @user is not activated
+      set_flash_message(:notice, :success, kind: 'Facebook') if is_navigational_format?
     else
-      session["devise.facebook_data"] = request.env["omniauth.auth"]
-      redirect_to new_user_registration_url
+      session['devise.facebook_data'] = auth
+      flash[:notice] = "Error: Your #{kind} account was not connected."
+      redirect_back fallback_location: new_user_session_path, allow_other_host: false
     end
   end
 
   def failure
-    redirect_to root_path
-    set_flash_message(:notice, :error, :kind => "Facebook") if is_navigational_format?
+    redirect_back fallback_location: new_user_session_path, allow_other_host: false
+    flash[:notice] = "Error: Your #{kind} account was not connected."
   end
 
+  def auth
+    request.env['omniauth.auth']
+  end
+
+  def set_user_service
+    return @user if defined? @user
+
+    if user_signed_in?
+      @user = current_user
+    elsif User.where(provider: auth.provider, uid: auth.uid).exists?
```

Comment:
From: https://stackoverflow.com/questions/11161663/find-vs-find-by-vs-where
It's more Ruby to use `find_by` if you are looking for one record in particular
Minor quibble as `find_by` becomes a `where`

Reply:
To be consistent, I would need to change all my `where` statements to `find_by` statements. I feel this is unnecessary work.